### PR TITLE
Clever/Powerschool multi-auth takeover: handle migrated users correctly

### PR DESCRIPTION
**Pull request by ewjordan**
Fixes improper handling of new migrated Clever and Powerschool users, so that account takeover works correctly.

### CODE REVIEW

File: dashboard/app/controllers/omniauth_callbacks_controller.rb [hunk 1]
```diff
@@ -145,9 +146,16 @@ def handle_untrusted_email_signin(user)
     else
       # Otherwise, it's either the first login, or a user who must connect -
       # offer to connect the Clever account to an existing one, or insist if needed
-      session['clever_link_flag'] = user.provider
-      session['clever_takeover_id'] = user.uid
-      session['clever_takeover_token'] = user.oauth_token
+      if user.migrated?
+        auth_option = authentication_options.find_by credential_type: provider
```
Comment (islemaster) [inline dashboard/app/controllers/omniauth_callbacks_controller.rb]: Where does `authentication_options` come from here?  Should it be `user.authentication_options.find_by ...`?

File: dashboard/app/helpers/users_helper.rb [hunk 2]
```diff
@@ -11,7 +11,7 @@ def check_and_apply_oauth_takeover(user)
     if session['clever_link_flag'].present? && session['clever_takeover_id'].present?
       uid = session['clever_takeover_id']
       # TODO: validate that we're not destroying an active account?
-      existing_account = User.where(uid: uid, provider: session['clever_link_flag']).first
+      existing_account = User.find_by_credential(type: session['clever_link_flag'], id: uid)
```
Comment (islemaster) [inline dashboard/app/helpers/users_helper.rb]: We use `session['clever_link_flag']` a couple of times in this function, it'd be nice to assign it to a `provider` variable up top like we do for `uid`.

File: dashboard/app/helpers/users_helper.rb [hunk 3]
```diff
@@ -21,10 +21,22 @@ def check_and_apply_oauth_takeover(user)
...
+      if user.migrated?
+        user.add_credential(
+          type: session['clever_link_flag'],
+          id: uid,
+          email: user.email,
+          hashed_email: user.hashed_email,
+          data: {
+            oauth_token: session['clever_takeover_token']
+          }
+        )
```
Comment (islemaster) [inline dashboard/app/helpers/users_helper.rb]: We're not doing anything with the return value of `add_credential` which should indicate whether the db interaction succeeded.  Should we make sure we raise or recover somehow if it fails?

File: dashboard/app/helpers/users_helper.rb [hunk 4]
```diff
@@ -21,10 +21,22 @@ def check_and_apply_oauth_takeover(user)
...
-      user.save
+      if user.migrated?
+        user.add_credential(
+          type: session['clever_link_flag'],
+          id: uid,
+          email: user.email,
+          hashed_email: user.hashed_email,
+          data: {
+            oauth_token: session['clever_takeover_token']
+          }
```
Comment (islemaster) [inline dashboard/app/helpers/users_helper.rb]: `AuthenticationOption#data` is still just a text field, and there's no clever before_save behavior to convert a hash to JSON (unless there's some rails magic going on I don't know about).  Do we need to add `.to_json` to the end of this hash?

File: dashboard/app/controllers/omniauth_callbacks_controller.rb (hunk 1 above)
Comment (ewjordan) [reply dashboard/app/controllers/omniauth_callbacks_controller.rb]: Ick...this is why I really need tests here, you're right, this doesn't work. Adding some.

File: dashboard/app/helpers/users_helper.rb (hunk 4 above)
Comment (ewjordan) [reply dashboard/app/helpers/users_helper.rb]: Very good catch.

Worse, this *will* successfully put the record into the database, in a *very* slightly different serialized format that's not quite JSON but looks kinda like it.

I found a couple other cases where we've done this in the code, fixed them up.

**Pull request merged**